src/ui/react/hooks.ts

```diff
@@ -19,7 +19,9 @@ export interface MassUpdater<Domain, K extends keyof Domain>
 type UpdaterInternal<T> = (options: {}, delta: (prev: T) => T) => void;
 export type Updater<T> = UpdaterInternal<T> & {
   (delta: (prev: T) => T): void;
-};
+} & (T extends {}
```

Comment:
just a question: So the correct way to define function overloads in TS is with `&` not `|`?

Reply:
The overloads even have their own syntax when you define them near the implementation. ([docs](https://www.typescriptlang.org/docs/handbook/functions.html#overloads)).

The distinction between `&` and `|` here is a matter of a correct perspective more broadly in types here. `|` is a union, but that means “OR” . `&` is interesection, but it means “AND”. In English, I’ll say this function does thingA _and_ thingB, which sounds like a union, but it’s an intersection. A union would be ”maybe it does thingA, but it could be doing thingB instead”.

Think about objects with fields: `{ name: string } & { email: string }` means the object has two fields. `{ name: string } | { email: string }` means we don’t know nothing for sure about the object.